Sources/SwiftProtobuf/SetPathDecoder.swift

```diff
@@ -0,0 +1,396 @@
+// Sources/SwiftProtobuf/SetPathDecoder.swift - Path decoder (Setter)
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Decoder which sets value of a field by its path.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+/// Describes errors can occure during decoding a proto by path.
+public enum PathDecodingError: Error {
+
+  /// Describes a mismatch in type of the fields.
+  ///
+  /// If a value of type A is applied to a path with type B.
+  /// this error will be thrown.
+  case typeMismatch
+
+  /// Describes path is not found in message type.
+  ///
+  /// If a message has no field with the given path this
+  /// error will be thrown.
+  case pathNotFound
+}
+
+extension Message {
+  static func number(for field: String) -> Int? {
+    guard let type = Self.self as? _ProtoNameProviding.Type else {
+      return nil
+    }
+    return type._protobuf_nameMap.number(forJSONName: field)
+  }
+}
+
+// Decoder that set value of a message field by the given path
+struct SetPathDecoder<T: Message>: Decoder {
+
+  // The value should be set to the path
+  private let value: Any?
+
+  // Field number should be overriden by decoder
+  private var number: Int?
+
+  // The path only including sub-paths
+  private let nextPath: [String]
+
+  // Merge options to be concidered while setting value
+  private let mergeOption: MergeOption
+
+  private var replaceRepeatedFields: Bool {
+    mergeOption.replaceRepeatedFields
+  }
+
+  init(
+    path: [String],
+    value: Any?,
+    mergeOption: MergeOption
+  ) {
+    if let firstComponent = path.first,
+       let number = T.number(for: firstComponent) {
+      self.number = number
+    }
+    self.nextPath = .init(path.dropFirst())
```

Comment:
Maybe also set `nextPath` to an empty array if the current component lookup fails?